### PR TITLE
bug fix http request pre closure

### PR DESCRIPTION
**Pull request by bishnuag**

### CODE REVIEW

File: runtime/client_http_request.go [hunk 1]
```diff
@@ -264,26 +264,35 @@ func (req *ClientHTTPRequest) executeDoWithRetry(ctx context.Context) (*http.Res
...
 
-//executeDo will send the request out with a timeout
+// executeDo will send the request out with a timeout
 func (req *ClientHTTPRequest) executeDo(ctx context.Context) (*http.Response, error) {
 	attemptCtx, cancelFn := context.WithTimeout(ctx, req.timeoutAndRetryOptions.RequestTimeoutPerAttemptInMs)
 	defer cancelFn()
-	return req.client.Client.Do(req.httpReq.WithContext(attemptCtx))
+	res, err := req.client.Client.Do(req.httpReq.WithContext(attemptCtx))
+	// when no error, read body and capture before closing the connection
+	if err != nil {
```
Comment (groshu) [inline runtime/client_http_request.go:288]: is err == nil ?

File: runtime/client_http_request.go [hunk 2]
```diff
@@ -264,26 +264,35 @@ func (req *ClientHTTPRequest) executeDoWithRetry(ctx context.Context) (*http.Res
...
 	return nil, retryCount, err
 }
 
-//executeDo will send the request out with a timeout
+// executeDo will send the request out with a timeout
 func (req *ClientHTTPRequest) executeDo(ctx context.Context) (*http.Response, error) {
 	attemptCtx, cancelFn := context.WithTimeout(ctx, req.timeoutAndRetryOptions.RequestTimeoutPerAttemptInMs)
 	defer cancelFn()
-	return req.client.Client.Do(req.httpReq.WithContext(attemptCtx))
+	res, err := req.client.Client.Do(req.httpReq.WithContext(attemptCtx))
```
Comment (groshu) [inline runtime/client_http_request.go:286]: can we add a small unit test for this ?

File: runtime/client_http_response.go [hunk 3]
```diff
@@ -66,8 +67,43 @@ func (res *ClientHTTPResponse) setRawHTTPResponse(httpRes *http.Response) {
 	res.Header = httpRes.Header
 }
 
+// readAndCaptureBody reads bytes from response and stores it in ClientHTTPResponse's rawResponseBytes
+func (res *ClientHTTPResponse) readAndCaptureBody() error {
+	// mark response read to true
+	defer func() {
+		res.responseRead = true
+	}()
+	rawBody, err := io.ReadAll(res.rawResponse.Body)
```
Comment (groshu) [inline runtime/client_http_response.go:76]: can we make this a fucntion will be able to capture the same logic in file below

File: runtime/client_http_request.go (hunk 1 above)
Comment (bishnuag) [reply runtime/client_http_request.go:288]: my bad 👍

File: runtime/client_http_request.go (hunk 2 above)
Comment (bishnuag) [reply runtime/client_http_request.go:286]: pls check TestMakingClientCallWithHeadersWithRequestLevelTimeoutAndRetries once

File: runtime/client_http_response.go (hunk 3 above)
Comment (bishnuag) [reply runtime/client_http_response.go]: updated the code, and made it more simple. Invoking ReadAll() method instead introducing new method

File: runtime/client_http_request.go [hunk 4]
```diff
@@ -264,26 +264,35 @@ func (req *ClientHTTPRequest) executeDoWithRetry(ctx context.Context) (*http.Res
...
-//executeDo will send the request out with a timeout
+// executeDo will send the request out with a timeout
 func (req *ClientHTTPRequest) executeDo(ctx context.Context) (*http.Response, error) {
-	attemptCtx, cancelFn := context.WithTimeout(ctx, req.timeoutAndRetryOptions.RequestTimeoutPerAttemptInMs)
-	defer cancelFn()
-	return req.client.Client.Do(req.httpReq.WithContext(attemptCtx))
+	ctx, cancel := context.WithTimeout(ctx, req.timeoutAndRetryOptions.RequestTimeoutPerAttemptInMs)
+	defer cancel()
+	res, err := req.client.Client.Do(req.httpReq.WithContext(ctx))
+	// when no error, read body and capture before closing the connection
```
Comment (hbhoyar-uber) [inline runtime/client_http_request.go:287]: nit: Add more context about context cancellation error if you read body outside the scope

**Review by hbhoyar-uber — APPROVED** [review]

**Review by groshu — APPROVED** [review]
LGTM

**Pull request merged**